### PR TITLE
KAFKA-7259: Remove deprecated ZKUtils usage from ZkSecurityMigrator

### PR DESCRIPTION
**Pull request by omkreddy**
- Remove ZKUtils usage from various tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by omkreddy** [comment]
@rajinisivaram @ijuma Please take a look, when you get a chance.

**Comment by harshach** [comment]
+1. LGTM.

**Review by satishd — COMMENTED** [review]
Thanks @omkreddy for the PR.

LGTM

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala [hunk 1]
```diff
@@ -120,17 +123,17 @@ object ZkSecurityMigrator extends Logging {
...
   private def setAcl(path: String, setPromise: Promise[String]) = {
     info("Setting ACL for path %s".format(path))
-    zkUtils.zkConnection.getZookeeper.setACL(path, zkUtils.defaultAcls(path), -1, SetACLCallback, setPromise)
+    zkClient.currentZooKeeper.setACL(path, zkClient.defaultAcls(path).asJava, -1, SetACLCallback, setPromise)
   }
 
   private def getChildren(path: String, childrenPromise: Promise[String]) = {
     info("Getting children to set ACLs for path %s".format(path))
-    zkUtils.zkConnection.getZookeeper.getChildren(path, false, GetChildrenCallback, childrenPromise)
+    zkClient.currentZooKeeper.getChildren(path, false, GetChildrenCallback, childrenPromise)
```
Comment (ijuma) [inline core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: `currentZooKeeper` is meant to be used for testing only.

Comment (omkreddy) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: @ijuma Thanks for the review. maybe we should allow  for this tool alone.

Other options are:
1. implementing similar behavior in KafkaZKClient methods
2. using ZooKeeper instance directly in ZkSecurityMigrator

What do you suggest?

**Comment by omkreddy** [comment]
@junrao Please take a look, when you get a chance.

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
retest this please

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 2]
```diff
@@ -1236,10 +1236,20 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   /**
    * Deletes the zk node recursively
    * @param path
+   * @param recursiveDelete
    * @return  return true if it succeeds, false otherwise
    */
-  def deletePath(path: String): Boolean = {
-    deleteRecursive(path)
+  def deletePath(path: String, recursiveDelete: Boolean = true): Boolean = {
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: None of the callers checks the return value. So, perhaps we can return nothing? Also, could we reuse this method in deleteLogDirEventNotifications(), deleteTopicDeletions(), and deletePartitionReassignment()?

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 3]
```diff
@@ -1433,6 +1443,32 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
...
+      case Code.OK => getAclResponse.acl
+      case _ => throw getAclResponse.resultException.get
+    }
+  }
+
+  /**
+    * sets the ACLs to the node of the given path
+    * @param path the given path for the node
+    * @param acl the given acl for the node
+    * @param aclVersion the given acl version of the node
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Should aclVersion be zkVersion?

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala [hunk 4]
```diff
@@ -538,6 +539,10 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
     zkClient.createRecursive(path)
     zkClient.deletePath(path)
     assertFalse(zkClient.pathExists(path))
+    zkClient.createRecursive(path)
+    zkClient.deletePath(path, false)
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala]: Should we also test delete /a recursively?

File: core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala [hunk 5]
```diff
@@ -19,14 +19,14 @@ package kafka
...
-
 import javax.imageio.ImageIO
+
 import kafka.admin.ReassignPartitionsCommand
 import kafka.admin.ReassignPartitionsCommand.Throttle
 import org.apache.kafka.common.TopicPartition
 import kafka.server.{KafkaConfig, KafkaServer, QuotaType}
 import kafka.utils.TestUtils._
-import kafka.utils.{Exit, Logging, TestUtils, ZkUtils}
+import kafka.utils._
```
Comment (junrao) [inline core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala]: Do we need every class in kafka.utils including ZkUtils?

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala [hunk 6]
```diff
@@ -52,8 +51,6 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslSetup {
    */
   @Test
   def testZkAclsDisabled() {
-    val zkUtils = ZkUtils(zkConnect, zkSessionTimeout, zkConnectionTimeout, zkAclsEnabled.getOrElse(JaasUtils.isZkSecurityEnabled))
-    TestUtils.verifyUnsecureZkAcls(zkUtils)
-    CoreUtils.swallow(zkUtils.close(), this)
+    TestUtils.verifyUnsecureZkAcls(zkClient)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:56]: Do we need to run TestUtils.verifyUnsecureZkAcls here and in SaslPlainSslEndToEndAuthorizationTest? Also, perhaps we can put that test in a class just for ZK testing.

**Review by junrao — COMMENTED** [review]
@omkreddy : Thanks for the patch. Just a few minor comments below.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: Personally, I'd replace this async style with the pipelined client. This is unnecessarily complex.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala (hunk 6 above)
Comment (omkreddy) [reply core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:56]: @junrao Thanks for the review.

we test ```TestUtils.verifySecureZkAcls``` in ```SaslPlainSslEndToEndAuthorizationTest```
and ```TestUtils.verifyUnsecureZkAcls``` in ```SaslPlainPlaintextConsumerTest```

First one is to verify  secure paths in ZK are created with the expected ACL and 
second one is to verify secure paths in ZK have no ACLs when zookeeper.set.acl=false and ZK SASL setup.

I have added a commit to enable ZooKeeper SASL in SaslPlainPlaintextConsumerTest.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala (hunk 1 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: Understood. If it's ok, I will do the change in another PR.

File: core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala (hunk 5 above)
Comment (omkreddy) [reply core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala]: Since ZkUtils.getReassignmentJson() method is only used here,  copied the method to  ReplicationQuotasTestRig.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 7]
```diff
@@ -1433,6 +1436,32 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
     }
   }
 
+  /**
+    * Return the ACLs of the node of the given path
+    * @param path the given path for the node
+    * @return the ACL array of the given node.
+    */
+  def getACL(path: String): Seq[ACL] = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: This is not following our naming convention, it should be getAcl.

File: core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala [hunk 8]
```diff
@@ -154,6 +154,19 @@ object ReplicationQuotasTestRig {
       println("Output can be found here: " + journal.path())
     }
 
+    def getReassignmentJson(partitionsToBeReassigned: Map[TopicPartition, Seq[Int]]): String = {
```
Comment (ijuma) [inline core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala]: Why do we need this? We should not adding code like this outside of the ZK classes.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: My main concern is that we are also widening the access for a number of methods that should only be accessible in the zk package. One option is to add a ZkSecurityMigratorUtils in the zk package that provides access to these methods while making it clear that they're not supposed to be used normally. And we can remove such a class when we migrate ZkSecurityMigrator away from ZK's asynchronous API.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala (hunk 6 above)
Comment (omkreddy) [reply core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:56]: @junrao Pls take a look at the updated PR

File: core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala (hunk 8 above)
Comment (omkreddy) [reply core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala]: This is a util method to convert Map to json string. This is only used in this test class. To avoid dependency on ZKutils class, I copied the method here. I can revert this change, later we can do this cleanup, when we remove Zkutils.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala (hunk 1 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: @ijuma  Thanks for the suggestion. Yes, this will make more clear to the developers. added ZkSecurityMigratorUtils class to to access zookeeper instance,

File: core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala (hunk 8 above)
Comment (ijuma) [reply core/src/test/scala/other/kafka/ReplicationQuotasTestRig.scala]: Yeah, I'd prefer to use `ZkUtils` than duplicate serialization logic.

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala [hunk 9]
```diff
@@ -1112,4 +1122,23 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
 
     assertEquals(expectedConsumerGroupOffsetsPath, actualConsumerGroupOffsetsPath)
   }
+
+  @Test
+  def testACLMethods(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala]: Our naming convention is `testAclMethods()`

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
@junrao  Pls take a look when you get a chance.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala [hunk 10]
```diff
@@ -120,17 +123,18 @@ object ZkSecurityMigrator extends Logging {
   }
 }
 
-class ZkSecurityMigrator(zkUtils: ZkUtils) extends Logging {
+class ZkSecurityMigrator(zkClient: KafkaZkClient) extends Logging {
+  private val zkSecurityMigratorUtils = new ZkSecurityMigratorUtils(zkClient)
   private val futures = new Queue[Future[String]]
 
   private def setAcl(path: String, setPromise: Promise[String]) = {
```
Comment (ryannedolan) [inline core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]:
```suggestion
  private def setAcl(path: String, setPromise: Promise[String]): Unit = {
```

Maybe fix these void methods while you're in here?

File: core/src/main/scala/kafka/zk/ZkSecurityMigratorUtils.scala [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+package kafka.zk
+
+import org.apache.zookeeper.ZooKeeper
+
+/**
+ * This class should only be used in ZkSecurityMigrator tool.
+ * This class will be removed after we migrate ZkSecurityMigrator away from ZK's asynchronous API.
+ * @param kafkaZkClient
+ */
+class ZkSecurityMigratorUtils(val kafkaZkClient: KafkaZkClient) {
```
Comment (ryannedolan) [inline core/src/main/scala/kafka/zk/ZkSecurityMigratorUtils.scala:26]: maybe private?

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala [hunk 12]
```diff
@@ -75,59 +81,76 @@ class ZkAuthorizationTest extends ZooKeeperTestHarness with Logging {
...
    * when isSecure is set to true.
    */
   @Test
-  def testZkUtils() {
-    assertTrue(zkUtils.isSecure)
-    for (path <- zkUtils.persistentZkPaths) {
-      zkUtils.makeSurePersistentPathExists(path)
-      if (ZkUtils.sensitivePath(path)) {
-        val aclList = zkUtils.zkConnection.getAcl(path).getKey
+  def testKafkaZkClient() {
```
Comment (ryannedolan) [inline core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala]:
```suggestion
  def testKafkaZkClient(): Unit {
```

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala [hunk 13]
```diff
@@ -75,59 +81,76 @@ class ZkAuthorizationTest extends ZooKeeperTestHarness with Logging {
...
+                               rack: Option[String] = None): BrokerInfo =
+    BrokerInfo(Broker(id, Seq(new EndPoint(host, port, ListenerName.forSecurityProtocol
+    (securityProtocol), securityProtocol)), rack = rack), ApiVersion.latestVersion, jmxPort = port + 10)
+
   /**
    * Tests the migration tool when making an unsecure
    * cluster secure.
    */
   @Test
   def testZkMigration() {
```
Comment (ryannedolan) [inline core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala]:
```suggestion
  def testZkMigration(): Unit {
```

File: core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala [hunk 14]
```diff
@@ -43,7 +43,7 @@ abstract class ZooKeeperTestHarness extends JUnitSuite with Logging {
   val zkSessionTimeout = 6000
   val zkMaxInFlightRequests = Int.MaxValue
 
-  protected val zkAclsEnabled: Option[Boolean] = None
+  protected def zkAclsEnabled: Option[Boolean] = None
```
Comment (ryannedolan) [inline core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala:46]: why this change? prefer val. If there is a problem with initialization order, consider lazy val.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala [hunk 15]
```diff
@@ -29,14 +28,16 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslSetup {
   private val kafkaServerJaasEntryName =
     s"${listenerName.value.toLowerCase(Locale.ROOT)}.${JaasTestUtils.KafkaServerContextName}"
   this.serverConfig.setProperty(KafkaConfig.ZkEnableSecureAclsProp, "false")
+  // disable secure acls of zkClient in ZooKeeperTestHarness
+  override protected def zkAclsEnabled = Some(false)
```
Comment (ryannedolan) [inline core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:32]: why def?

**Review by ryannedolan — COMMENTED** [review]
couple scala style nits

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala (hunk 10 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: Updated the void methods, though we are not strictly following this style

File: core/src/main/scala/kafka/zk/ZkSecurityMigratorUtils.scala (hunk 11 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/zk/ZkSecurityMigratorUtils.scala:26]: This class is accessed in other package

File: core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala (hunk 14 above)
Comment (omkreddy) [reply core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala:46]: I need to initialize this value in one of the classes. made this def so that we can override in subclasses.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala (hunk 15 above)
Comment (omkreddy) [reply core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:32]: This is related above comment. This is where I need to override the zkAclsEnabled.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 16]
```diff
@@ -538,11 +538,8 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
    * @param expectedControllerEpochZkVersion expected controller epoch zkVersion.
    */
   def deleteLogDirEventNotifications(sequenceNumbers: Seq[String], expectedControllerEpochZkVersion: Int): Unit = {
-    val deleteRequests = sequenceNumbers.map { sequenceNumber =>
-      DeleteRequest(LogDirEventNotificationSequenceZNode.path(sequenceNumber), ZkVersion.MatchAnyVersion,
-        zkVersionCheck = controllerZkVersionCheck(expectedControllerEpochZkVersion))
-    }
-    retryRequestsUntilConnected(deleteRequests)
+    sequenceNumbers.foreach(sequenceNumber => deletePath(LogDirEventNotificationSequenceZNode.path(sequenceNumber),
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Performance wise, it's still better to process all the deletes in a pipelined way as before.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 17]
```diff
@@ -761,9 +758,7 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
    * @param expectedControllerEpochZkVersion expected controller epoch zkVersion.
    */
   def deleteTopicDeletions(topics: Seq[String], expectedControllerEpochZkVersion: Int): Unit = {
-    val deleteRequests = topics.map(topic => DeleteRequest(DeleteTopicsTopicZNode.path(topic), ZkVersion.MatchAnyVersion,
-      zkVersionCheck = controllerZkVersionCheck(expectedControllerEpochZkVersion)))
-    retryRequestsUntilConnected(deleteRequests)
+    topics.foreach(topic => deletePath(DeleteTopicsTopicZNode.path(topic), expectedControllerEpochZkVersion))
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Performance wise, it's still better to process all the deletes in a pipelined way as before.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 18]
```diff
@@ -1241,11 +1234,21 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
...
    * Deletes the zk node recursively
-   * @param path
+   * @param path path to delete
+   * @param expectedControllerEpochZkVersion expected controller epoch zkVersion.
+   * @param recursiveDelete enable recursive delete
    * @return  return true if it succeeds, false otherwise
    */
-  def deletePath(path: String): Boolean = {
-    deleteRecursive(path)
+  def deletePath(path: String, expectedControllerEpochZkVersion: Int = ZkVersion.MatchAnyVersion, recursiveDelete: Boolean = true): Unit = {
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala:1247]: The comment on return value needs to be adjusted.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 19]
```diff
@@ -1439,6 +1442,32 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
...
+    }
+  }
+
+  /**
+    * sets the ACLs to the node of the given path
+    * @param path the given path for the node
+    * @param acl the given acl for the node
+    * @param zkVersion the given zk version of the node
+    */
+  def setAcl(path: String, acl: Seq[ACL], zkVersion: Int): Unit = {
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Hmm, do we need zkVersion? It doesn't seem that it's used in SetAclRequest.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala [hunk 20]
```diff
@@ -29,14 +28,16 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslSetup {
...
   override protected def securityProtocol = SecurityProtocol.SASL_PLAINTEXT
   override protected lazy val trustStoreFile = Some(File.createTempFile("truststore", ".jks"))
   override protected val serverSaslProperties = Some(kafkaServerSaslProperties(kafkaServerSaslMechanisms, kafkaClientSaslMechanism))
   override protected val clientSaslProperties = Some(kafkaClientSaslProperties(kafkaClientSaslMechanism))
 
   @Before
   override def setUp(): Unit = {
-    startSasl(jaasSections(kafkaServerSaslMechanisms, Some(kafkaClientSaslMechanism), KafkaSasl, kafkaServerJaasEntryName))
+    startSasl(jaasSections(kafkaServerSaslMechanisms, Some(kafkaClientSaslMechanism), Both, kafkaServerJaasEntryName))
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:40]: Why do we need to change KafkaSasl to Both?

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala [hunk 21]
```diff
@@ -75,59 +81,76 @@ class ZkAuthorizationTest extends ZooKeeperTestHarness with Logging {
...
     }
-    // Test that can create: createEphemeralPathExpectConflict
-    zkUtils.createEphemeralPathExpectConflict("/a", "")
-    verify("/a")
-    // Test that can create: createPersistentPath
-    zkUtils.createPersistentPath("/b")
-    verify("/b")
+
+    // Test that creates Ephemeral node
+    val brokerInfo = createBrokerInfo(1, "test.host", 9999, SecurityProtocol.PLAINTEXT)
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala:107]: It seems that most of the added tests in this method are already covered in KafkaZkClientTest? If not, they probably should be moved over to KafkaZkClientTest?

**Review by junrao — COMMENTED** [review]
@omkreddy : Thanks for the updated patch. Added a few more comments.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala (hunk 16 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Agree. reverted this change for performance reasons.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala (hunk 17 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Agree. reverted this change for performance reasons.

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala (hunk 19 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Yes, we are not using now. removed the version parameter. We can add if needed in future

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala (hunk 20 above)
Comment (omkreddy) [reply core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:40]: setting ```Both``` will enable kerberos for both Kafka and ZK.  Previously we were [testing](https://github.com/omkreddy/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala#L50) without enabling ZK SASL. 
Now we enabled ZK sasl and and test that everyone can access ZkUtils.SecureZkRootPaths/ ZkUtils.SensitiveZkRootPaths when zookeeper.set.acl=false

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala (hunk 21 above)
Comment (omkreddy) [reply core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala:107]: These tests are to verify the created ZK ACLs using KafkaZKClient. We are just reusing the kerberized zk setup available in this class for this test. 

Current KafkaZkClientTest tests run on non-kerberised zk. Let me know, if we want to create separate test class for this.

File: core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala (hunk 20 above)
Comment (junrao) [reply core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:40]: thanks

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala (hunk 21 above)
Comment (junrao) [reply core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala:107]: sounds good

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala [hunk 22]
```diff
@@ -136,118 +159,118 @@ class ZkAuthorizationTest extends ZooKeeperTestHarness with Logging {
...
   }
 
   /**
    * Verifies that the path has the appropriate secure ACL.
    */
   private def verify(path: String): Boolean = {
-    val sensitive = ZkUtils.sensitivePath(path)
-    val list = zkUtils.zkConnection.getAcl(path).getKey
-    list.asScala.forall(TestUtils.isAclSecure(_, sensitive))
+    val sensitive = ZkData.sensitivePath(path)
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala:259]: This is an existing issue. It seems that we should either assert in verify or assert the return value of verify?

**Review by junrao — COMMENTED** [review]
@omkreddy : Thanks for the updated patch. Just one more comment below.

File: core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala (hunk 22 above)
Comment (omkreddy) [reply core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala:259]: @junrao  Thanks for the catch!  added a assert statement. Updated the PR.

**Review by junrao — APPROVED** [review]
@omkreddy : Thanks for the latest update. LGTM. I will let you merge this.

**Comment by omkreddy** [comment]
@junrao  Thanks for all the reviews.
